packages/flutter/lib/src/widgets/async.dart

```diff
@@ -0,0 +1,387 @@
+/// The data and error fields of snapshots produced are only changed when the
+/// state is `ConnectionState.active`.
+///
+/// See also:
+///
+/// * [StreamBuilderBase], which supports widget building based on a computation
+/// that spans all interactions made with the stream.
+class StreamBuilder<T> extends StreamBuilderBase<T, AsyncSnapshot<T>> {
+  /// Creates a new [StreamBuilder] that builds itself based on the latest
+  /// snapshot of interaction with the specified [stream] and whose build
+  /// strategy is given by [builder].
+  StreamBuilder({
+    Key key,
+    Stream<T> stream,
+    @required this.builder
+  }) : super(key: key, stream: stream) {
+    assert(builder != null);
+  }
+
+  /// The build strategy currently used by this builder. Cannot be `null`.
+  final AsyncWidgetBuilder<T> builder;
+
+  @override
+  AsyncSnapshot<T> initial() => new AsyncSnapshot<T>.nothing();
+
+  @override
+  AsyncSnapshot<T> afterConnected(AsyncSnapshot<T> current) => current.inState(ConnectionState.waiting);
+
+  @override
+  AsyncSnapshot<T> afterData(AsyncSnapshot<T> current, T data) => new AsyncSnapshot<T>.activeData(data);
+
+  @override
+  AsyncSnapshot<T> afterError(AsyncSnapshot<T> current, Object error) => new AsyncSnapshot<T>.activeError(error);
+
+  @override
+  AsyncSnapshot<T> afterDone(AsyncSnapshot<T> current) => current.inState(ConnectionState.done);
+
+  @override
+  AsyncSnapshot<T> afterDisconnected(AsyncSnapshot<T> current) => current.inState(ConnectionState.none);
+
+  @override
+  Widget build(BuildContext context, AsyncSnapshot<T> currentSummary) => builder(context, currentSummary);
+}
+
+/// Widget that builds itself based on the latest snapshot of interaction with
+/// a [Future].
+///
+/// Behaves identically to [StreamBuilder] configured with `future?.asStream()`.
+class FutureBuilder<T> extends StatefulWidget {
+  FutureBuilder({
+    Key key,
+    this.future,
+    @required this.builder
+  }) : super(key: key) {
+    assert(builder != null);
+  }
+
+  /// The asynchronous computation to which this builder is currently connected,
+  /// possibly `null`.
+  final Future<T> future;
+
+  /// The build strategy currently used by this builder. Cannot be `null`.
+  final AsyncWidgetBuilder<T> builder;
+
+  @override
+  State<FutureBuilder<T>> createState() => new _FutureBuilderState<T>();
+}
+
+/// State for [FutureBuilder].
+class _FutureBuilderState<T> extends State<FutureBuilder<T>> {
+  Stream<T> _stream;
+
+  @override
+  void initState() {
+    super.initState();
+    _subscribe();
+  }
+
+  @override
+  void didUpdateConfig(FutureBuilder<T> oldConfig) {
+    if (oldConfig.future != config.future) {
+      _unsubscribe();
+      _subscribe();
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) => new StreamBuilder<T>(stream: _stream, builder: config.builder);
+
+  @override
+  void dispose() {
+    _unsubscribe();
+    super.dispose();
+  }
+
+  void _subscribe() {
+    _stream = config.future?.asStream();
```

Comment:
It turns out the re-implementation using the Future API has a minor deviation from the Stream-based one. With streams, the '''build''' method can (at least in theory) be called after the Future's completion and before the stream's done event has been consumed, thus resulting in '''build''' being called with a snapshot in '''ConnectionState.active'''. That cannot happen with the Future-based implementation which goes directly to '''ConnectionState.done''' when the Future completes.

The difference is unlikely to be seen in practice because all known implementations of '''Future.asStream()''' actually produce the completion and the done event in the same loop of the event mill (according to lrn@). Thus the re-implementation avoids the creation of additional AsyncSnapshots that are not used at all.